Announce opam 2.2.0~alpha3 (#96)
ocaml/platform-blog: opam-2-2-0-alpha3.md

title: "opam 2.2.0 alpha3 release"
authors: [
  "Raja Boujbel - OCamlPro" {"mailto:raja.boujbel(à)ocamlpro.com"}
  "Kate Deplaix - Tarides" {"mailto:kit-ty-kate(à)outlook.com"}
  "David Allsopp - Tarides" {"mailto:david(à)tarides.com"}
]
date: "2023-11-13"
--BODY--

<!--
_Feedback on this post is welcomed on [Discuss](https://discuss.ocaml.org/t/ann-opam-2-2-0-alpha3-release/12699)!_
---!>

We are happy to announce the third alpha release of opam 2.2.0. It contains
some fixes and a new feature for Windows.  You can view the full list in the
[release note](https://github.com/ocaml/opam/releases/tag/2.2.0-alpha3).

This version is an alpha, we invite users to test it to spot previously
unnoticed bugs as we head towards the stable release.

## Major change: Environment variables handling on Windows

> The `x-env-path-rewrite` extensions field is now handled specially.

Reviewer: Possible alternative (intending the markdown to be taken from the comment, if wanted!):

opam files now support a new `x-env-path-rewrite` field which
specifies rewriting rules for the environment variable updates defined in
the `setenv` and `build-env` fields. This field allows greater control over the
separator character used for PATH-like fields (i.e. `;` vs `:`), conversion of
slashes to backslashes, and even conversion from Windows native path format
(`C:\cygwin64\home\dra\.opam`) to Cygwin format (`/home/dra/.opam`).

The rewriting rules allow opam directory-like variables (e.g. the `%{lib}%` directory
of a switch) to be used in `setenv` and `build-env` fields in a portable way.

For example, given[^1]:
```
setenv: [
  [PKG_CONFIG_PATH += "%{lib}%/pkgconfig"]
  [PATH += "%{share}%/bin"]
]
x-env-path-rewrite: [
  [ PKG_CONFIG_PATH ":" "host" ]
  [ PATH (":" {os != "win32"} | ";" {os = "win32"}) ("target" {os != "win32"} | "target-quoted" {os = "win32"}) ]
]
```
the environment variable changes given by `opam env` on Windows would be:

```
PKG_CONFIG_PATH='/cygdrive/c/Users/DRA/AppData/Local/opam/default/lib/pkgconfig[:<rest-of-PKG_CONFIG_PATH, if given>]'
PATH='C:\Users\DRA\AppData\Local\opam\default\share\bin;C:\Users\DRA\AppData\Local\opam\default\bin;<rest-of-PATH>'
```

with the following interesting parts for Windows users:
- `PKG_CONFIG_PATH`, which is consumed by a Cygwin-tool, has the directory given in Unix-like syntax, and opam's `share` variable was automatically converted
- The correct separator is used for each (`:` for `PKG_CONFIG_PATH`, `;` for `PATH` is used when adding entries)
- In the `PATH` update, `/bin` was converted to `\bin`

The full syntax is described in full in the [manual](https://opam.ocaml.org/doc/Manual.html#env-update-rewrite).

Opam files carrying this new field are still compatible with the opam 2.0 syntax as it is an extension field, however its effect is only available with opam 2.2.0~alpha3 and above.
If you want to make sure users of the package containing it have a compatible opam, you can use the `available` field to that end:
```
available: opam-version >= "2.2.0~alpha3"
```
or, if the change is Windows-specific:
```
available: opam-version >= "2.2.0~alpha3" | os != "win32"
```

[^1]: Note that the specification for `PATH` is opam's _default_ behaviour, so it's not actually necessary to have this formula for `PATH` in the `x-env-path-rewrite` field

Author: Integrated in https://github.com/ocaml/platform-blog/pull/96/commits/175e99d6a13354458e1042263aa391f8be3800f2

```
setenv: [
  [PKG_CONFIG_PATH += "%{lib}%/pkgconfig"]
  [PATH += "%{share}%/bin"]
]
x-env-path-rewrite: [
  [ PKG_CONFIG_PATH ":" "host" ]
  [ PATH (":" {os != "win32"} | ";" {os = "win32"}) ("target" {os != "win32"} | "target-quoted" {os = "win32"}) ]
]
```

In this example `PKG_CONFIG_PATH` will be a list separated by `:` and the added path will always be in its unix form (on Windows `cygpath` will be used).
`PATH` will be separated by `:` on Unix-like systems and `;` on Windows. The path will also be transformed to use backslashes on Windows and quoted if the path contains semi-colons.

The full syntax is described in full in the [manual](https://opam.ocaml.org/doc/Manual.html#env-update-rewrite).

Opam files carrying this new field are still compatible with the opam 2.0 syntax as it is an extension field, however its effect is only available with opam 2.2.0~alpha3 and above.
If you want to make sure users of the package containing it have a compatible opam, you can use the `available` field to that end:
```
available: opam-version >= "2.2.0~alpha3"
```
if the change is Windows specific, you can use:
```
available: opam-version >= "2.2.0~alpha3" | os != "win32"
```

> ## Other note-worthy changes
> 
> * Sandbox: Make /tmp writable again to restore POSIX compliancy
> * opam tree: Allow packages with a specific version, directories or local opam files, as input
> * opam tree: Handle `--recurse` and `--subpath` for directory arguments
> * opam admin: Add the `add-extrafiles` command to add/check/update the `extra-files:` field according to the files present in the `files/` directory
> * opam lint: Allow to mark a set of warnings as errors using a new syntax -W @1..9
> * Releases: Pre-built binaries now include ppc64le and s390x

Reviewer: I'd switch the voice for changes to say what opam now does (`/tmp` is now writeable) vs what was done to opam (Make [opam make] `/tmp` writable again...)
```suggestion
## Other noteworthy changes

* Sandbox: `/tmp` is now writable again, restoring POSIX compliance
* opam tree: `opam tree package.version` is now supported, displaying the dependency tree of a specific version of a package
* opam tree: `--recurse` and `--subpath` are supported for directory arguments
* opam admin: new `add-extrafiles` command to add/check/update the `extra-files:` field according to the files present in the `files/` directory
* opam lint: new syntax allow marking a set of warnings as errors e.g. `-W @1..9`
* Releases: Pre-built binaries now include ppc64le and s390x
```

## Miscellaneous changes

> * A handful of issues related to the compilation of opam on Windows, were fixed

Reviewer: nit:
```suggestion
* A handful of issues related to the compilation of opam on Windows were fixed
```

* Bugs in the handling of the `OPAMCURL`, `OPAMFETCH` and `OPAMVERBOSE` environment variables were fixed
* Bugs in the handling of the `--assume-built` argument were fixed
* Sporadic crashes and segfaults during shell detection on Windows were fixed

> * The switch specific `environment` file now stores environment variable rewriting rules

Reviewer: Is this final entry "worth" it? (are there any users/systems which actually read our `environment` files?)

Author: I was debating keeping it or not, but thinking about it again i think it doesn't make sense for this blog post, you're right.
```suggestion
```

Various other improvements were made and bugs were fixed.
API changes are denoted in the release note linked above.

> This release also includes a handful of PRs improving the documentation and more than a dozen PRs improving an extending the tests.

Reviewer: typo:
```suggestion
This release also includes a handful of PRs improving the documentation and more than a dozen PRs improving and extending the tests.
```

## Windows Support

The main opam-repository Windows compliance is still a work in progress, we
recommend to use existing [compatible
repository](https://github.com/ocaml-opam/opam-repository-mingw) (originally
from [@fdopen](https://github.com/fdopen)) and [32/64 bit mingw-w64

> packages](https://github.com/dra27/opam-repository) (by

Reviewer: ```suggestion
packages](https://github.com/dra27/opam-repository/tree/windows-5.0) (by
```

Otherwise you can compile opam yourself using the following steps:

This alpha requires a preexisting Cygwin installation for compiling opam.

1. Check that you have all dependencies installed:
  * `autoconf`, `make`, `patch`, `curl`
  * MinGW compilers: `mingw64-x86_64-gcc-g++`, `mingw64-i686-gcc-g++`
  * Or if you want to use the MSVC port of OCaml, you'll need to install Visual Studio or Visual Studio Build Tools
2. Download & extract the [opam archive](https://github.com/ocaml/opam/releases/download/2.2.0-alpha3/opam-full-2.2.0-alpha3.tar.gz)
3. In the directory:
  a. if you are using MSVC: launch `make cold`
  b. if you are using MinGW: make sure the path to the `libc` dlls are in your `PATH` and launch `make cold`. For instance: `export PATH='C:\cygwin64\usr\x86_64-w64-mingw32\sys-root\mingw\bin':$PATH && make cold`. Don’t forget to update `PATH` accordingly or place the `opam` binary in the same directory as the `libc` dlls if you want to move the resulting opam binary.

>   c. alternatively, if you're using MinGW: `make cold CONFIGURE_ARGS=--with-private-runtime`. If you change the location of the resulting opam binary, don't forget to copy `Opam.Runtime.amd64` (or `Opam.Runtime.i386`) in the same directory.

Reviewer 2: ```suggestion
  c. alternatively, if you're using MinGW: `make cold CONFIGURE_ARGS=--with-private-runtime`. If you change the location of the resulting opam binary, don't forget to copy `Opam.Runtime.amd64` directory (or `Opam.Runtime.i386`) in the same directory.
```

4. A coffee later, you now have an opam executable!

> 5. Start your preferred Windows terminal (`cmd` or `PowerShell`), and initialise opam with the Windows _sunset_ repository:

Reviewer 2: should we re-add here that path need to be update if mingw without private runtime ?

Author: I don’t think it’s necessary at this point given it’s been said 2 lines above

title: "opam 2.2.0 alpha3 release"
authors: [
  "Raja Boujbel - OCamlPro" {"mailto:raja.boujbel(à)ocamlpro.com"}
  "Kate Deplaix - Tarides" {"mailto:kit-ty-kate(à)outlook.com"}
  "David Allsopp - Tarides" {"mailto:david(à)tarides.com"}
]
date: "2023-11-13"
--BODY--

<!--
_Feedback on this post is welcomed on [Discuss](https://discuss.ocaml.org/t/ann-opam-2-2-0-alpha3-release/12699)!_

> ---!>

Reviewer 2: ```suggestion
--!>
```

* opam lint: new syntax allow marking a set of warnings as errors e.g. `-W @1..9`
* Releases: Pre-built binaries now include ppc64le and s390x

## Miscellaneous changes

* A handful of issues related to the compilation of opam on Windows were fixed
* Bugs in the handling of the `OPAMCURL`, `OPAMFETCH` and `OPAMVERBOSE` environment variables were fixed
* Bugs in the handling of the `--assume-built` argument were fixed
* Sporadic crashes and segfaults during shell detection on Windows were fixed

Various other improvements were made and bugs were fixed.
API changes are denoted in the release note linked above.
This release also includes a handful of PRs improving the documentation and more than a dozen PRs improving and extending the tests.

## Windows Support

The main opam-repository Windows compliance is still a work in progress, we
recommend to use existing [compatible
repository](https://github.com/ocaml-opam/opam-repository-mingw) (originally
from [@fdopen](https://github.com/fdopen)) and [32/64 bit mingw-w64
packages](https://github.com/dra27/opam-repository/tree/windows-5.0) (by
[@dra27](https://github.com/dra27)).

### How to Test opam on Windows

> 

Author: ```suggestion
should we mention the experimental prebuilt windows binary?
```

Reviewer 2: yes! for testing purpose with the experimental flag

Author: ```suggestion
If you're feeling adventurous, you can use the experimental pre-built binary for Windows available [here](https://github.com/ocaml/opam/releases/download/2.2.0-alpha3/opam-2.2.0-alpha3-preview-for-windows.exe). It should work but since it's our first public pre-built binary for Windows please use at your own risk.

Otherwise you can compile opam yourself using the following steps:

```

Reviewer 2: i like it :)

recommend to use existing [compatible
repository](https://github.com/ocaml-opam/opam-repository-mingw) (originally
from [@fdopen](https://github.com/fdopen)) and [32/64 bit mingw-w64
packages](https://github.com/dra27/opam-repository/tree/windows-5.0) (by
[@dra27](https://github.com/dra27)).

### How to Test opam on Windows
If you're feeling adventurous, you can use the experimental pre-built binary for Windows available [here](https://github.com/ocaml/opam/releases/download/2.2.0-alpha3/opam-2.2.0-alpha3-preview-for-windows.exe). It should work but since it's our first public pre-built binary for Windows please use at your own risk.

Otherwise you can compile opam yourself using the following steps:

This alpha requires a preexisting Cygwin installation for compiling opam.

1. Check that you have all dependencies installed:
  * `autoconf`, `make`, `patch`, `curl`
  * MinGW compilers: `mingw64-x86_64-gcc-g++`, `mingw64-i686-gcc-g++`
  * Or if you want to use the MSVC port of OCaml, you'll need to install Visual Studio or Visual Studio Build Tools
2. Download & extract the [opam archive](https://github.com/ocaml/opam/releases/download/2.2.0-alpha3/opam-full-2.2.0-alpha3.tar.gz)

> 3. In the directory launch `make cold`

Author: ```suggestion
3. In the directory:
  a. if you are using MSVC: launch `make cold`
  b. if you are using MinGW: launch either make sure the path to the `libc` dlls are in your `PATH` or use `make cold PATH='C:\cygwin64\usr\x86_64-w64-mingw32\sys-root\mingw\bin':$PATH` or `make cold CONFIGURE_ARGS=--with-private-runtime`, and in the later case, if you change the location of the resulting opam binary, don't forget to copy `Opam.Runtime.amd64` (or `Opam.Runtime.i386`) with it.
```

Reviewer 3: I like the change, clearer.

Reviewer 3: The part on Opam.Runtime.xxx is still very confusing though. Is it a folder? Where do we find this thing and why should we care?

Reviewer 2: It is a directory generated by `make cold` with `--with-private-runtime`, that contains the dlls.

Reviewer 2: note: the path update is not only needed for compilation, but also for launching opam. It needs to be persistent.

4. A coffee later, you now have an opam executable!
5. Start your preferred Windows terminal (`cmd` or `PowerShell`), and initialise opam with the Windows _sunset_ repository:
  * `opam init git+https://github.com/ocaml-opam/opam-repository-mingw`

From here, you can try to install the [sunset
repository](https://discuss.ocaml.org/t/sunsetting-opam-repository-mingw/11632)
packages. If you find any bugs, please [submit an
issue](https://github.com/ocaml-opam/opam-repository-mingw#what-do-i-do-when-things-are-broken).
It will help opam-repository maintainers to add Windows repository packages
into the main repository.

> > Hint: if you use the MinGW compiler, don't forget to add to your `PATH` the
> path to `libc` dlls (usually
> `C:\cygwin64\usr\x86_64-w64-mingw32\sys-root\mingw\bin`). Or compile opam with
> `make cold CONFIGURE_ARGS=--with-private-runtime`, and if you change opam
> location, don't forget to copy `Opam.Runtime.amd64` (or `Opam.Runtime.i386`)
> with it.
> 

Author: ```suggestion
```
